### PR TITLE
Use std filesystem

### PR DESCRIPTION
**Pull request by OznOg**

### CODE REVIEW

**Comment by Baklap4** [comment]
Looks promising already :) Keep up the work! :)

**Comment by OznOg** [comment]
Compilation fails because filesystem support with gcc 7 and clang 5 is still experimental.
That said, both are >2 years old now, is there a possiblility to move to >gcc8 and >clang6 ?
The other option is adding more compilation flags to add 'experimental' in header and namespace, but I'm affraid this makes code really painful to read.
Could be also that I add extra compiler check and do not activate the feature for older compiler, but ther the CMakeList.txt will get some more combinatory on compilers.

Let me know what do you prefer...

**Comment by Tronic** [comment]
Drop the old versions.

**Comment by Baklap4** [comment]
Drop the old version, let's get some new stuff in.

**Comment by Lord-Kamina** [comment]
I'd like to keep support for now but definitely think we should drop them eventually.

**Comment by Baklap4** [comment]
but when is 'eventually'? in most projects i worked on 'eventually' equals never... While in this code why not clean it up?

**Comment by Tronic** [comment]
The big question is, does anyone still use those old versions? Even Debian stable has GCC 8 and Clang 7.

**Comment by Lord-Kamina** [comment]
If somebody is still on Ubuntu 16, they would.

**Comment by OznOg** [comment]
> If somebody is still on Ubuntu 16, they would.

Are people on ubuntu 16.04 compiling performous and/or expecting new packages considering that even the compiler/stdlib support for c++17 is partial? IMHO, people with ubuntu 16 are welcome to use older packages of performous, I don't think they expect backport of the bleeding edge.

**Comment by Baklap4** [comment]
> Are people on ubuntu 16.04 compiling performous 

Currently the semaphore ci is still building it, aside from that i think most developers did indeed upgrade to 17.04 or 18.04 already, that's like 2 years ago so...

**Comment by Baklap4** [comment]
@OznOg can we make this up to date with latest master?
It'd be nice to support this

**Comment by OznOg** [comment]
so what are the takeaways here, do we drop support for ubuntu16.04 (now 20.04 is released) ? do we move to completely to c++17?

**Comment by Lord-Kamina** [comment]
Note: macOS does support `<filesystem>` from Catalina onwards.

**Comment by OznOg** [comment]
Do you know how to check properly the different versions of MacOS stl?

My journey is not over, seems that ubuntu also complains...

**Comment by OznOg** [comment]
legacy compilers keeps failing, do I really need to try a fix? clang 5 and gcc7 are quite old....

**Comment by Lord-Kamina** [comment]
Both of them should be fully c++17 compliant, though. So, it's likely the issue is with the flags being passed to them.

**Comment by Lord-Kamina** [comment]
> Both of them should be fully c++17 compliant, though. So, it's likely the issue is with the flags being passed to them.

I see I'm evidently wrong; despite they supporting the standard it seems they don't support the `<filesystem>` header. Anyway, this does not still necessary mean the end for xenial; using gcc-8 is a drop-in replacement for gcc-7 on our CI, as the PPAs we're currently using already include it and I'm sure something similar can be done about Clang.

**Comment by OznOg** [comment]
or we can just say we use boost for compilers that do not have filesystem's support

**Comment by Lord-Kamina** [comment]
> or we can just say we use boost for compilers that do not have filesystem's support

I don't think performous as it currently is builds on vanilla 16.04 anyway. However, with the PPAs the CI is currently using, both clang-6 and gcc-8 are available so it's not really an issue.

**Comment by OznOg** [comment]
ok, great :), I need to do some cleanup and we are ready for reviews

**Comment by Lord-Kamina** [comment]
For the record, this command in terminal will give you the macOS version number:
`/usr/bin/sw_vers | grep ProductVersion | cut -f2 -d: | cut -f2 -d"    "`

**Comment by OznOg** [comment]
do you have any update on this?

**Comment by OznOg** [comment]
FYI, I tested this using strace and logged all files accessed with boost version and with c++17. I was able to fix a few things with this mean; now outputs are matching exactly.

**Comment by OznOg** [comment]
any update?

**Comment by OznOg** [comment]
please give me some update on "merge or not". If you don't feel like merging it, let's just drop it, but I don't like stalled branches for ever.

**Comment by Baklap4** [comment]
eey Oznog, sorry for not getting back to you earlier.
I had some issues regarding my pc and getting the parts of it delivered.
Anytime this week i'll install linux on it and test this merge request as i'd love to have this within performous.

**Comment by OznOg** [comment]
@Baklap4 any update? is it worth keeping boost compat now we are on 2021 and FS support is 4 years old?

**Comment by OznOg** [comment]
hum... ping?

**Comment by OznOg** [comment]
ping

File: CMakeLists.txt [hunk 1]
```diff
@@ -30,25 +34,33 @@ if (USE_BOOST_REGEX)
...
-	message("Detected GCC version ${GCC_VERSION}")
-	if (GCC_VERSION VERSION_GREATER 7 OR GCC_VERSION VERSION_EQUAL 7)
-		set(CMAKE_CXX_FLAGS "-std=c++14 -Wall -Wextra -fcx-limited-range ${CMAKE_CXX_FLAGS}")
-	elseif (GCC_VERSION VERSION_GREATER 4.8 OR GCC_VERSION VERSION_EQUAL 4.8)
-		set(CMAKE_CXX_FLAGS "-std=c++1y -Wall -Wextra -Wno-attributes -Wno-deprecated-declarations -fcx-limited-range ${CMAKE_CXX_FLAGS}")
-		message(WARNING "Your GCC is rather old, disabling some useful compiler warnings (GCC 7+ or Clang 5+ recommended for proper C++17 support).")
+if (APPLE)
+	set (MACOS_SUPPORT_FILESYSTEM FALSE)
+	if (${CMAKE_OSX_DEPLOYMENT_TARGET})
+		message (STATUS "Chcking macOS deployment target... ${CMAKE_OSX_DEPLOYMENT_TARGET}")
```
Comment (Baklap4) [inline CMakeLists.txt:42]: Typo: "Chcking" -> "Checking"

**Review by Baklap4 — DISMISSED** [review]
Would be nice to fix this typo, but aside from that i couldn't find things and it's working alright.
Did anyone @performous/core-owners  else end up having problems?

In regard of boost compat.. there's another issue to clean up boost dependencies, let's fix that there? (https://github.com/performous/performous/issues/39)

File: CMakeLists.txt (hunk 1 above)
Comment (OznOg) [reply CMakeLists.txt]: done

Comment (OznOg) [reply CMakeLists.txt]: done

**Review by Baklap4 — APPROVED** [review]
LGTM

**Comment by OznOg** [comment]
:champagne:

**Pull request merged**